cdbot/cogs/admin.py

```diff
@@ -0,0 +1,55 @@
+import re
+from discord import Member
+from discord.ext.commands import Bot, Cog, Context, command
+
+from cbot.constants import (
+    PLACEHOLDER_NICKNAME, NICKNAME_PATTERNS
+)
+
+def checkName(nick):
+    result = False
+    for i in NICKNAME_PATTERNS:
+        if re.match(i, nick):
+            result = True
+            break
+
+    return result
+
+
+class Admin(Cog):
+    """
+    Admin functionality
+    """
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    @Cog.listener()  # triggered on new/removed nickname
+    async def on_member_update(memberBefore, memberAfter):
+        newNickname = memberAfter.display_name
+        if checkName(newNickname):  # bad display name
+            if not checkName(memberAfter.name):  # username is okay
+                await memberAfter.edit(nick=None)  # reset nickname
+            else:
+                # assign placeholder nickname
+                await memberAfter.edit(nick=PLACEHOLDER_NICKNAME)
+    
+
+    @Cog.listener()  # triggered on username change
+    async def on_user_update(memberBefore, memberAfter):
+        newUsername = memberAfter.name
+        if checkName(newUsername):  # bad username
+            # assign placeholder nickname
+            await memberAfter.edit(nick=PLACEHOLDER_NICKNAME)
+
+
+    @Cog.listener()
+    async def on_member_join(member):
+        username = member.name
+        if checkName(username):  # bad username
+            # assign placeholder nickname
+            await member.edit(nick=PLACEHOLDER_NICKNAME)
```

Comment:
Have you considered resetting the nickname entirely (to the username) instead of to a predefined constant? If so, what's the reasoning out of interest?

Reply:
I believe the code at this line is checking if the username of the user joining is suitable; setting a placeholder nickname for the user removes the offending username of the newly joined user. I believe the idea is to reset the nickname if the username is OK, otherwise set the nickname of the user to hide the username and remove the nickname.

Reply:
That users could have inappropriate usernames, or usernames that cannot be typed on an English keyboard (due to using non-ASCII/non-Latin characters).

Reply:
Furthermore, if the username is fine, it will reset the nickname to the username.

Reply:
Ah, yep, nevermind